Feature/댓글 crud#35

rudeh1253 opened the pull request:
# 이슈번호 
#35 

# 개요  
- Member의 memberId를 String 타입에서 Long 타입으로 변경 (테이블 스키마도 같이 변경)
- 인증/인가 로직에 memberId 데이터 타입 변경 반영
- 댓글 entity 정의
- 댓글 관련 CRUD 기능 구현
- 댓글 관련 Access Control 구현
- 회원 API URI에 /hr prefix 덧붙임

# 추가사항
변경 사항이 너무 많네요 죄송합니다 ㅠㅠ 다음부터는 풀 리퀘 나눠서 올리겠습니다

rudeh1253 committed:
feat: Comment entity 정의

rudeh1253 committed:
refactor: 테이블 스키마 변경에 따른 코드 변경

memberId를 Long형으로 변경하고 loginId 칼럼 추가. 이에 따른 회원관리 및
인증/인가 코드 수정 및 Swagger docs 수정

rudeh1253 committed:
feat: 댓글 작성 기능 구현

주어진 postId, writerId에 따라 댓글 작성

rudeh1253 committed:
feat: 댓글 삭제

Soft Delete 적용

rudeh1253 committed:
feat: Post별 댓글 조회 기능 구현

rudeh1253 committed:
feat: 회원별 댓글 목록 조회 기능 구현

rudeh1253 committed:
feat: 댓글 수정 기능 구현

rudeh1253 committed:
feat: 댓글 관련 exception handling 및 Swagger 명세 작성

rudeh1253 committed:
refactor: 회원 API URI에 /hr prefix 추가

rudeh1253 committed:
feat: Comment API에 대한 AccessController 정의

rudeh1253 committed:
refactor: 회원가입 API request body 롤백

기존 API를 사용하던 프론트엔드의 수정을 (조금이라도) 줄이기 위해 기존에
사용하던 request body 구조로 롤백

rudeh1253 committed:
refactor: 인증/인가 로직에 Long형 memberId가 사용되도록 변경

rudeh1253 committed:
refactor: Comment에서 BaseTimeEntity 상속

rudeh1253 committed:
refactor: 통합 과정에서 에러가 나는 부분 수정

memberId가 String 타입에서 Long 타입으로 바뀜에 따라 발생한 에러들 해결

Related to: #35

dlrkdus commented on src/main/java/com/hf/healthfriend/domain/comment/controller/CommentController.java:
```diff
@@ -0,0 +1,104 @@
+    public ResponseEntity<ApiBasicResponse<CommentCreationResponseDto>> createComment(
+            @PathVariable("postId") long postId,
+            @RequestBody CommentCreationRequestDto requestDto) {
+        CommentCreationResponseDto result = this.commentService.createComment(postId, requestDto);
```
이렇게 주입한 서비스 앞에 this 키워드를 사용하는 이유가 왜인가요??

rudeh1253 replied:
로컬 변수와 구분을 두기 위해 필드를 사용할 때는 습관적으로 this 키워드를 붙이고 있습니다.
제 개인적인 습관인 부분이 큽니다.

dlrkdus commented on src/main/java/com/hf/healthfriend/domain/comment/dto/response/CommentCreationResponseDto.java:
```diff
@@ -0,0 +1,20 @@
+@Builder
+@Getter
+@ToString
+public class CommentCreationResponseDto {
```
저는 Response 타입 DTO는 record를 선호합니다. 응답값은 불변인 경우가 많으니까요. 취향차일 순 있겠습니다 ㅎㅎ

rudeh1253 replied:
record는 많이 사용해 보지 않아서 getter 메소드명이 get~ 형식이 아닐 때 reflection을 사용하는 서드 파티 라이브러리에서 어떻게 받아들일지 확신이 안 서서 잘 안 써봤습니다.
이번 기회에 한번 사용해 보겠습니다! 조언 감사합니다!

dlrkdus commented on src/main/java/com/hf/healthfriend/domain/comment/entity/Comment.java:
```diff
@@ -0,0 +1,65 @@
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
+@Builder
+@Getter
+@Setter
```
엔티티에 Setter를 사용한 이유가 따로 있으신가요?

rudeh1253 replied:
엔티티 업데이트를 위해 Setter를 사용했습니다.
지금까지 엔티티 업데이트를 위해 엔티티에 보통 Setter를 사용했는데 가연님께서 의문을 가져 주신 덕에 엔티티에 Setter를 추가하는 걸 지양해야 하는 이유를 찾아볼 수 있었습니다. 레퍼런스를 찾아서 엔티티를 수정하는 더 안전한 방법을 찾을 수 있었습니다.
덕분에 더 좋은 practice를 알아갑니다! 감사합니다!

dlrkdus commented on src/main/java/com/hf/healthfriend/domain/comment/repository/CommentRepository.java:
```diff
@@ -0,0 +1,71 @@
+            throw new CommentNotFoundException("Comment entity of commentId not exists", commentId);
+        }
+        Comment comment = commentOp.get();
+        comment.setDeleted(true);
```
setter보단 메소드를 사용하는게 (ex `delete()` ) 유지보수나 명시적인 면에서도 낫다고 생각합니다.

rudeh1253 replied:
제가 안일하게 Setter를 정의한 탓에 delete 메소드 대신에 setter를 사용한 것 같습니다. 좋은 지적 감사합니다!
이 부분도 리팩토링해 놓겠습니다.

dlrkdus commented on src/main/java/com/hf/healthfriend/domain/comment/dto/request/CommentCreationRequestDto.java:
```diff
@@ -0,0 +1,14 @@
+@ToString
+public class CommentCreationRequestDto {
+    private long writerId;
+    private String content;
```
기본적인 Validation 들어가면 좋을 것 같습니다 !

rudeh1253 replied:
옙 추가하겠습니다!

dlrkdus commented on src/main/java/com/hf/healthfriend/domain/comment/repository/CommentRepository.java:
```diff
@@ -25,16 +26,15 @@ public Comment save(Comment commentToSave) {
-    public boolean deleteById(Long commentId) {
+    public void deleteById(Long commentId) throws CommentNotFoundException {
         Optional<Comment> commentOp = this.commentJpaRepository.findById(commentId);
         if (commentOp.isEmpty()) {
```
comment 객체가 없을 때만 예외가 던져지고 있습니다. 논리 삭제이기 때문에 isDeleted가 true일 때도 예외를 던져줘야 합니다.

rudeh1253 replied:
아 이 점은 놓치고 있었던 부분입니다. 지적 감사합니다!
isDeleted가 true일 경우 아예 레코드를 안 가져오도록 CommentJpaRepository에서 findById 메소드를 오버라이딩하겠습니다!

rudeh1253 committed:
refactor(comment): 코드 리뷰 반영

- 엔티티에 Setter 제거
- BaseTimeEntity의 setter의 접근자를 protected로 변경
- CommentCreationResponseDto에 record 적용
- Setter를 사용한 코드 리팩토링

Related to: #35

rudeh1253 reviewed (commented):
좋은 리뷰 감사합니다! 덕분에 배워갈 수 있었습니다!

dlrkdus reviewed (approved):
LGTM 👍
